### PR TITLE
feat: Add bounds to raster source

### PR DESCRIPTION
**Pull request by Elter71**
## Description

Added `bounds support for RasterSource` . Now, you can limit the bounds within which raster tiles will render.

## Checklist

- [X] I've read `CONTRIBUTING.md`
- [x] I updated the doc/other generated code with running `yarn generate` in the root folder
- [x] I have tested the new feature on `/example` app.
  - [x] In V11 mode/ios
  - [x] In New Architecture mode/ios
  - [x] In V11 mode/android
  - [x] In New Architecture mode/android
- [X] I added/updated a sample - if a new feature was implemented (`/example`)

## Screenshot OR Video
Screenshot_1730212439
Simulator Screenshot - iPhone SE (3rd generation) - 2024-10-29 at 15 34 47

### CODE REVIEW

File: android/src/main/java/com/rnmapbox/rnmbx/components/styles/sources/RNMBXRasterSourceManager.kt [hunk 1]
```diff
@@ -42,4 +46,36 @@ class RNMBXRasterSourceManager(reactApplicationContext: ReactApplicationContext)
...
     }
-}
\ No newline at end of file
+
+    @ReactProp(name = "sourceBounds")
+    override fun setSourceBounds(source: RNMBXRasterSource, value: Dynamic) {
+        if (value.type.name == "Array") {
+            val readableArray: ReadableArray = value.asArray()
+
+            if (readableArray.size() == 4) {
```
Comment (mfazekas) [inline android/src/main/java/com/rnmapbox/rnmbx/components/styles/sources/RNMBXRasterSourceManager.kt:55]: Is this check necessary here? We'll just silently error if the size is not 4

File: android/src/main/java/com/rnmapbox/rnmbx/components/styles/sources/RNMBXTileSource.kt [hunk 2]
```diff
@@ -35,10 +36,14 @@ abstract class RNMBXTileSource<T : Source?>(context: Context?) : RNMBXSource<T>(
         if (attribution != null) {
             builder.attribution(attribution)
         }
+        if(bounds != null) {
+             val boundsArray = bounds!!.clone()
```
Comment (mfazekas) [inline android/src/main/java/com/rnmapbox/rnmbx/components/styles/sources/RNMBXTileSource.kt:40]: do we need `!!` we've tested above for null already

**Review by mfazekas — CHANGES_REQUESTED** [review]
@Elter71 Thanks much amazing work!

Some minor nitpicks

File: android/src/main/java/com/rnmapbox/rnmbx/components/styles/sources/RNMBXTileSource.kt (hunk 2 above)
Comment (Elter71) [reply android/src/main/java/com/rnmapbox/rnmbx/components/styles/sources/RNMBXTileSource.kt:40]: Without this, you will get an error:

`Smart cast to 'Array<Double>' is impossible, because 'bounds' is a mutable property that could have been changed by this time FAILURE: Build failed with an exception.`

File: android/src/main/java/com/rnmapbox/rnmbx/components/styles/sources/RNMBXRasterSourceManager.kt (hunk 1 above)
Comment (Elter71) [reply android/src/main/java/com/rnmapbox/rnmbx/components/styles/sources/RNMBXRasterSourceManager.kt:55]: Without this, you will get an error like this:
Screenshot_1730282753

Personally, I prefer a more direct message, but I leave it up to you.

File: android/src/main/java/com/rnmapbox/rnmbx/components/styles/sources/RNMBXTileSource.kt (hunk 2 above)
Comment (mfazekas) [reply android/src/main/java/com/rnmapbox/rnmbx/components/styles/sources/RNMBXTileSource.kt:40]: We should be able to do something like this:
```kt
bounds?.let { 
   builder.bounds(Arrays.asList(it))
}
```

File: android/src/main/java/com/rnmapbox/rnmbx/components/styles/sources/RNMBXRasterSourceManager.kt (hunk 1 above)
Comment (mfazekas) [reply android/src/main/java/com/rnmapbox/rnmbx/components/styles/sources/RNMBXRasterSourceManager.kt:55]: Sorry, mybad, you're right I've missed the Logger.e at the end. 

Nitpick: Can please you rewrite the logic so we have early return on error statements, and not, the success case 

Something like:
```swift
if (value.type.name != "Array" || readableArray.size() != 4) {
   Logger.e(RNMBXRasterSourceManager.REACT_CLASS, "source bounds must be an array with left, bottom, top, and right values")
   return
}

val bboxArray = Array(4) { i -> readableArray.getDouble(i)
source.setSourceBounds(bboxArray)
...

**Comment by Elter71** [comment]
@mfazekas All suggestions have been applied. Thanks for the quick response!

**Review by mfazekas — APPROVED** [review]

**Comment by mfazekas** [comment]
@Elter71 thanks much, amazing!
